src/Adapter/Grid/Data/Factory/CartGridDataFactory.php

```diff
@@ -0,0 +1,147 @@
+use PrestaShop\PrestaShop\Core\Grid\Data\Factory\GridDataFactoryInterface;
+use PrestaShop\PrestaShop\Core\Grid\Data\GridData;
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollection;
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use PrestaShop\PrestaShop\Core\Localization\LocaleInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Decorates Doctrine data factory for carts
+ */
+final class CartGridDataFactory implements GridDataFactoryInterface
+{
+    /**
+     * @var GridDataFactoryInterface
+     */
+    private $doctrineDataFactory;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var LocaleInterface
+     */
+    private $locale;
+
+    /**
+     * @param GridDataFactoryInterface $doctrineDataFactory
+     * @param TranslatorInterface $translator
+     * @param LocaleInterface $locale
+     */
+    public function __construct(
+        GridDataFactoryInterface $doctrineDataFactory,
+        TranslatorInterface $translator,
+        LocaleInterface $locale
+    ) {
+        $this->doctrineDataFactory = $doctrineDataFactory;
+        $this->translator = $translator;
+        $this->locale = $locale;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData(SearchCriteriaInterface $searchCriteria)
+    {
+        $data = $this->doctrineDataFactory->getData($searchCriteria);
+
+        $records = $this->applyModifications($data->getRecords());
+
+        return new GridData($records, $data->getRecordsTotal(), $data->getQuery());
+    }
+
+    /**
+     * @param RecordCollectionInterface $records
+     *
+     * @return RecordCollection
+     *
+     * @throws CartException
+     */
+    private function applyModifications(RecordCollectionInterface $records)
+    {
+        $modifiedRecords = [];
+
+        foreach ($records as $record) {
+            switch ($record['status']) {
+                case 'abandoned_cart':
+                    $record['status'] = $this->translator->trans(
+                        'Abandoned cart',
+                        [],
+                        'Admin.Orderscustomers.Feature'
+                    );
+                    break;
+                case 'not_ordered':
+                    $record['status'] = $this->translator->trans('Not placed', [], 'Admin.Orderscustomers.Feature');
+                    break;
+            }
+
+            if (empty($record['carrier_name'])) {
+                $record['carrier_name'] = '--';
+            }
+
+            if (empty($record['customer_name'])) {
+                $record['customer_name'] = '--';
+            }
+
+            $record['online'] = $record['id_guest'] ?
+                $this->translator->trans('Yes', [], 'Admin.Global') :
+                $this->translator->trans('No', [], 'Admin.Global');
+
+            $context = Context::getContext();
```

Comment:
As we needed to do this nasty stuff multiple times, we created `ContextStateManager` to handle it. Please use ContextStateManager to make all this horrible code happen in a single place 😅

Reply:
done